File: packages/hushspec/src/observer.ts
import type { EvaluationAction, EvaluationResult, Decision } from './evaluate.js';
import type { DecisionReceipt, EnforcementSummary } from './receipt.js';
import type { HushSpec } from './schema.js';
import { evaluateWithDetection } from './detection.js';

export interface EvaluationEvent {
  type:
    | 'evaluation.completed'
    | 'policy.loaded'
    | 'policy.load_failed'
    | 'policy.reloaded'
    | 'sink.error'
    | 'error';
  timestamp: string;
}

export interface EvaluationCompletedEvent extends EvaluationEvent {
  type: 'evaluation.completed';
  /** The action with `content` stripped (receipt spec 4.4). */
  action: EvaluationAction;
  /**
   * True when `action.content` was present and removed. The flag sits on the
   * event rather than on the action because `EvaluationAction` is a closed
   * wire type -- an SDK reading the event back rejects a member it does not
   * declare.
   */
  content_redacted?: boolean;
  result: EvaluationResult;
  duration_us: number;
  receipt?: DecisionReceipt;
  enforcement?: EnforcementSummary;
}

export interface PolicyLoadedEvent extends EvaluationEvent {
  type: 'policy.loaded';
  policy_name?: string;
  /**
   * The canonical content hash of the resolved policy (`sha256:` + hex), the
   * same value a receipt's `policy.content_hash` carries. Empty only when the
   * emitter had no policy to hash.
   */
  content_hash: string;
}

export interface PolicyLoadFailedEvent extends EvaluationEvent {
  type: 'policy.load_failed';
  error: string;
  source?: string;
}

export interface PolicyReloadedEvent extends EvaluationEvent {
  type: 'policy.reloaded';
  policy_name?: string;
  /** @see {@link PolicyLoadedEvent.content_hash} */
  content_hash: string;
  previous_hash?: string;
}

/**
 * A receipt sink refused what it was handed. The decision it belonged to
 * stands: a sink is evidence, never enforcement.
 */
export interface SinkErrorEvent extends EvaluationEvent {
  type: 'sink.error';
  error: string;
  /** The sink that refused, by name. */
  source?: string;
}

/**
 * A failure the guard absorbed that is neither a policy load nor a sink: an
 * observer that threw, say. The evaluation that produced the event stands --
 * an observer is a bystander, never enforcement.
 */
export interface ObserverErrorEvent extends EvaluationEvent {
  type: 'error';
  error: string;
  /** What failed, when the emitter can name it. */
  source?: string;
}

export type ObserverEvent =
  | EvaluationCompletedEvent
  | PolicyLoadedEvent
  | PolicyLoadFailedEvent
  | PolicyReloadedEvent
  | SinkErrorEvent
  | ObserverErrorEvent;

export interface EvaluationObserver {
  onEvent(event: ObserverEvent): void;
}

export class JsonLineObserver implements EvaluationObserver {
  constructor(private stream: NodeJS.WritableStream) {}

  onEvent(event: ObserverEvent): void {
    this.stream.write(JSON.stringify(event) + '\n');
  }
}

export class ConsoleObserver implements EvaluationObserver {
  constructor(private level: 'all' | 'deny_only' = 'all') {}

  onEvent(event: ObserverEvent): void {
    if (
      this.level === 'deny_only' &&
      event.type === 'evaluation.completed' &&
      event.result.decision !== 'deny'
    ) {
      return;
    }
    console.error(`[hushspec] ${event.type} at ${event.timestamp}`, event);
  }
}

/**
 * Upper bounds of the latency histogram, in microseconds. The last bucket is
 * `+Inf`, which the exposition adds.
 */
export const DURATION_BUCKETS_US: readonly number[] = [
  10, 25, 50, 100, 250, 500, 1_000, 5_000, 10_000,
];

/**
 * How many recent durations a {@link MetricsCollector} keeps for its
 * percentile. A collector lives as long as the process, so the window is
 * bounded; the counters and the histogram it reports are exact regardless.
 */
export const DURATION_WINDOW = 10_000;

/** One action type's latency histogram. */
interface Histogram {
  /** Cumulative counts, aligned with {@link DURATION_BUCKETS_US}. */
  buckets: number[];
  sum: number;
  count: number;
}

/**
 * Counters and a latency histogram over the evaluation stream, rendered as
 * Prometheus text.
 *
 * The exposed series are the ones the observability specification names, so a
 * dashboard or a recording rule works against any SDK unchanged:
 *
 * | Series | Type | Labels |
 * |---|---|---|
 * | `hushspec_evaluate_total` | counter | `decision`, `action_type` |
 * | `hushspec_evaluate_duration_us` | histogram | `action_type` |
 * | `hushspec_rule_match_total` | counter | `rule_block`, `decision` |
 * | `hushspec_policy_load_total` | counter | `status` |
 */
export class MetricsCollector implements EvaluationObserver {
  private counts: Map<string, number> = new Map();
  /** `decision` and `action_type` -> count. */
  private evaluations: Map<string, number> = new Map();
  /** `rule_block` and `decision` -> count. */
  private ruleMatches: Map<string, number> = new Map();
  /** `success` / `failure` -> count. */
  private policyLoads: Map<string, number> = new Map();
  /** `action_type` -> histogram. */
  private histograms: Map<string, Histogram> = new Map();
  /** A bounded window of recent samples, for the percentile. */
  private durations: number[] = [];
  private durationsAt = 0;
  private evaluationCount = 0;

  constructor(private readonly durationWindow: number = DURATION_WINDOW) {}

  onEvent(event: ObserverEvent): void {
    if (event.type === 'evaluation.completed') {
      this.recordEvaluation(event);
    } else if (event.type === 'policy.loaded' || event.type === 'policy.reloaded') {
      increment(this.policyLoads, 'success');
    } else if (event.type === 'policy.load_failed') {
      increment(this.policyLoads, 'failure');
    }
    increment(this.counts, event.type);
  }

  private recordEvaluation(event: EvaluationCompletedEvent): void {
    const decision = event.result.decision;
    const actionType = event.action.type;
    increment(this.counts, `evaluate.${decision}`);
    increment(this.evaluations, labelKey(decision, actionType));

    const ruleBlock = ruleBlockOf(event.result.matched_rule);
    if (ruleBlock !== undefined) {
      increment(this.ruleMatches, labelKey(ruleBlock, decision));
    }

    const durationUs = Math.max(0, event.duration_us);
    let histogram = this.histograms.get(actionType);
    if (histogram === undefined) {
      histogram = { buckets: DURATION_BUCKETS_US.map(() => 0), sum: 0, count: 0 };
      this.histograms.set(actionType, histogram);
    }
    histogram.count += 1;
    histogram.sum += durationUs;
    DURATION_BUCKETS_US.forEach((bound, index) => {
      if (durationUs <= bound) histogram.buckets[index] += 1;
    });

    this.evaluationCount += 1;
    this.recordSample(durationUs);
  }

  /** Keep the newest `durationWindow` samples, overwriting the oldest. */
  private recordSample(durationUs: number): void {
    if (this.durationWindow < 1) return;
    if (this.durations.length < this.durationWindow) {
      this.durations.push(durationUs);
      return;
    }
    this.durations[this.durationsAt] = durationUs;
    this.durationsAt = (this.durationsAt + 1) % this.durationWindow;
  }

  getCount(key: string): number {
    return this.counts.get(key) ?? 0;
  }

  getTotalEvaluations(): number {
    return this.evaluationCount;
  }

  /** Mean latency over the sample window, in microseconds. */
  getAverageDurationUs(): number {
    if (this.durations.length === 0) return 0;
    return this.durations.reduce((a, b) => a + b, 0) / this.durations.length;
  }

  /** 99th percentile latency over the sample window, in microseconds. */
  getP99DurationUs(): number {
    if (this.durations.length === 0) return 0;
    const sorted = [...this.durations].sort((a, b) => a - b);
    // `floor(n * 0.99) < n` for every n >= 1, so the index is always in range.
    return sorted[Math.floor(sorted.length * 0.99)];
  }

  /** Prometheus text exposition (version 0.0.4) of every series. */
  toPrometheus(): string {
    const lines: string[] = [];

    lines.push('# HELP hushspec_evaluate_total Total HushSpec evaluations');
    lines.push('# TYPE hushspec_evaluate_total counter');
    for (const [pair, count] of sortedEntries(this.evaluations)) {
      const [decision, actionType] = splitKey(pair);
      lines.push(
        `hushspec_evaluate_total{decision="${decision}",action_type="${escapeLabel(actionType)}"} ${count}`,
      );
    }

    lines.push('# HELP hushspec_evaluate_duration_us Evaluation duration in microseconds');
    lines.push('# TYPE hushspec_evaluate_duration_us histogram');
    for (const [actionType, histogram] of [...this.histograms].sort(byKey)) {
      const label = escapeLabel(actionType);
      DURATION_BUCKETS_US.forEach((bound, index) => {
        lines.push(
          `hushspec_evaluate_duration_us_bucket{action_type="${label}",le="${bound}"} ${histogram.buckets[index]}`,
        );
      });
      lines.push(
        `hushspec_evaluate_duration_us_bucket{action_type="${label}",le="+Inf"} ${histogram.count}`,
      );
      lines.push(`hushspec_evaluate_duration_us_sum{action_type="${label}"} ${histogram.sum}`);
      lines.push(`hushspec_evaluate_duration_us_count{action_type="${label}"} ${histogram.count}`);
    }

    lines.push('# HELP hushspec_rule_match_total Rule block match counts');
    lines.push('# TYPE hushspec_rule_match_total counter');
    for (const [pair, count] of sortedEntries(this.ruleMatches)) {
      const [ruleBlock, decision] = splitKey(pair);
      lines.push(
        `hushspec_rule_match_total{rule_block="${escapeLabel(ruleBlock)}",decision="${decision}"} ${count}`,
      );
    }

    lines.push('# HELP hushspec_policy_load_total Policy load operations');
    lines.push('# TYPE hushspec_policy_load_total counter');
    for (const [status, count] of sortedEntries(this.policyLoads)) {
      lines.push(`hushspec_policy_load_total{status="${status}"} ${count}`);
    }

    return lines.join('\n') + '\n';
  }

  reset(): void {
    this.counts.clear();
    this.evaluations.clear();
    this.ruleMatches.clear();
    this.policyLoads.clear();
    this.histograms.clear();
    this.durations = [];
    this.durationsAt = 0;
    this.evaluationCount = 0;
  }
}

/** Two label values as one map key; NUL cannot occur in either. */
function labelKey(first: string, second: string): string {
  return `${first}\0${second}`;
}

function splitKey(pair: string): [string, string] {
  const index = pair.indexOf('\0');
  return [pair.slice(0, index), pair.slice(index + 1)];
}

function increment(counts: Map<string, number>, at: string): void {
  counts.set(at, (counts.get(at) ?? 0) + 1);
}

function byKey(left: [string, unknown], right: [string, unknown]): number {
  return left[0] < right[0] ? -1 : left[0] > right[0] ? 1 : 0;
}

function sortedEntries(counts: Map<string, number>): [string, number][] {
  return [...counts].sort(byKey);
}

function escapeLabel(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
}

/**
 * The rule block a `matched_rule` belongs to, for the
 * `hushspec_rule_match_total` label.
 *
 * `rules.egress.default` -> `egress`; `extensions.posture.budgets` ->
 * `posture`; the bare `detection` the detection pipeline emits -> `detection`;
 * a reserved `__hushspec_x__` id -> `hushspec_x`. `undefined` for an
 * evaluation no rule decided (a default allow).
 */
export function ruleBlockOf(matchedRule?: string): string | undefined {
  if (matchedRule === undefined || matchedRule === '') return undefined;
  if (matchedRule.startsWith('__')) return matchedRule.replace(/^_+|_+$/g, '');
  if (matchedRule.startsWith('rules.')) return firstSegment(matchedRule.slice('rules.'.length));
  if (matchedRule.startsWith('extensions.')) {
    return firstSegment(matchedRule.slice('extensions.'.length));
  }
  return firstSegment(matchedRule);
}

function firstSegment(path: string): string {
  const index = path.search(/[.[]/);
  return index < 0 ? path : path.slice(0, index);
}

export class ObservableEvaluator {
  private observers: EvaluationObserver[] = [];

  addObserver(observer: EvaluationObserver): void {
    this.observers.push(observer);
  }

  removeObserver(observer: EvaluationObserver): void {
    this.observers = this.observers.filter(o => o !== observer);
  }

  /**
   * Evaluate `action` against `spec`, time it, and announce the outcome.
   *
   * Routes through the detection pipeline, as every other evaluation surface
   * does: for a policy carrying an `extensions.detection` block an escalated
   * decision must not come back as an allow simply because the caller went
   * through an observer (detection spec section 4).
   */
  evaluate(spec: HushSpec, action: EvaluationAction): EvaluationResult {
    const start = performance.now();
    const result = evaluateWithDetection(spec, action).evaluation;
    const duration_us = Math.round((performance.now() - start) * 1000);
    this.emit({
      type: 'evaluation.completed',
      timestamp: new Date().toISOString(),
      action,
      result,
      duration_us,
    });
    return result;
  }

  notifyEvaluationCompleted(
    action: EvaluationAction,
    result: EvaluationResult,
    durationUs: number,
    enforcement?: EnforcementSummary,
    receipt?: DecisionReceipt,
  ): void {
    this.emit({
      type: 'evaluation.completed',
      timestamp: new Date().toISOString(),
      action,
      result,
      duration_us: durationUs,
      enforcement,
      receipt,
    });
  }

  notifyPolicyLoaded(name?: string, hash?: string): void {
    this.emit({
      type: 'policy.loaded',
      timestamp: new Date().toISOString(),
      policy_name: name,
      content_hash: hash ?? '',
    });
  }

  notifyPolicyLoadFailed(error: string, source?: string): void {
    this.emit({
      type: 'policy.load_failed',
      timestamp: new Date().toISOString(),
      error,
      source,
    });
  }

  notifySinkError(error: string, source?: string): void {
    this.emit({
      type: 'sink.error',
      timestamp: new Date().toISOString(),
      error,
      source,
    });
  }

  notifyPolicyReloaded(name?: string, hash?: string, previousHash?: string): void {
    this.emit({
      type: 'policy.reloaded',
      timestamp: new Date().toISOString(),
      policy_name: name,
      content_hash: hash ?? '',
      previous_hash: previousHash,
    });
  }

  /** Announce a failure the guard absorbed that is neither a load nor a sink. */
  notifyError(error: string, source?: string): void {
    this.emit({
      type: 'error',
      timestamp: new Date().toISOString(),
      error,
      source,
    });
  }

  /**
   * Fan one event out, absorbing a throw.
   *
   * An observer is a bystander: it never decides whether an action proceeds,
   * so a throw in one must stop neither the evaluation that produced the event
   * nor the rest of the fan-out. The failure is handed back to the same
   * observer as an `error` event, because a failure nobody is told about is
   * the one that goes unnoticed; an observer that throws reporting its own
   * throw is dropped rather than retried.
   *
   * An `evaluation.completed` event is redacted here, so no path to an
   * observer can carry an action's `content` (receipt spec 4.4).
   */
  private emit(event: ObserverEvent): void {
    if (event.type === 'evaluation.completed') {
      event = redactCompleted(event);
    }
    for (const observer of this.observers) {
      try {
        observer.onEvent(event);
      } catch (thrown) {
        if (event.type === 'error') continue;
        const message = thrown instanceof Error ? thrown.message : String(thrown);
        try {
          observer.onEvent({
            type: 'error',
            timestamp: new Date().toISOString(),
            error: `observer threw: ${message}`,
          });
        } catch {
          /* best-effort: an observer that cannot be told is left alone */
        }
      }
    }
  }
}

/**
 * The event with `action.content` stripped, flagged on the event itself.
 *
 * Evidence records an action's content by hash and size, never by its bytes
 * (receipt spec 4.4), and an observer stream is held to the same rule.
 */
function redactCompleted(event: EvaluationCompletedEvent): EvaluationCompletedEvent {
  if (event.action.content == null) return event;
  const { content: _content, ...action } = event.action;
  return { ...event, action, content_redacted: true };
}

File: packages/python/hushspec/observer.py
from __future__ import annotations

import dataclasses
import enum
import json
import sys
import threading
import time
from abc import ABC, abstractmethod
from collections import deque
from datetime import datetime, timezone
from typing import Any, Optional, TextIO
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from hushspec.receipt import DecisionReceipt, EnforcementSummary

from hushspec.evaluate import EvaluationAction, EvaluationResult
from hushspec.schema import HushSpec

#: How many recent durations :class:`MetricsCollector` keeps for its
#: percentile. A collector lives as long as the process, so the window is
#: bounded; the counters and the histogram it reports are exact regardless.
DURATION_WINDOW = 10_000

#: Upper bounds of the latency histogram, in microseconds. The last bucket is
#: ``+Inf``, which the exposition adds.
DURATION_BUCKETS_US: tuple[int, ...] = (10, 25, 50, 100, 250, 500, 1_000, 5_000, 10_000)


class _Histogram:
    """One action type's cumulative latency histogram."""

    __slots__ = ("buckets", "total", "count")

    def __init__(self) -> None:
        self.buckets = [0] * len(DURATION_BUCKETS_US)
        self.total = 0.0
        self.count = 0

    def record(self, duration_us: float) -> None:
        self.count += 1
        self.total += duration_us
        for index, bound in enumerate(DURATION_BUCKETS_US):
            if duration_us <= bound:
                self.buckets[index] += 1

    def copy(self) -> "_Histogram":
        clone = _Histogram()
        clone.buckets = list(self.buckets)
        clone.total = self.total
        clone.count = self.count
        return clone


def rule_block_of(matched_rule: Optional[str]) -> Optional[str]:
    """The rule block a ``matched_rule`` belongs to, for the metric label.

    ``rules.egress.default`` is ``egress``; ``extensions.posture.budgets`` is
    ``posture``; the bare ``detection`` the detection pipeline emits is
    ``detection``; a reserved ``__hushspec_x__`` id is ``hushspec_x``. ``None``
    for an evaluation no rule decided (a default allow).
    """
    if not matched_rule:
        return None
    if matched_rule.startswith("__"):
        return matched_rule.strip("_")
    for prefix in ("rules.", "extensions."):
        if matched_rule.startswith(prefix):
            return _first_segment(matched_rule[len(prefix):])
    return _first_segment(matched_rule)


def _first_segment(path: str) -> str:
    for index, character in enumerate(path):
        if character in ".[":
            return path[:index]
    return path


def _escape_label(value: str) -> str:
    return value.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n")


def _format_number(value: float) -> str:
    """A duration sum as Prometheus reads it: an integer when it is whole."""
    return str(int(value)) if value == int(value) else repr(value)


class EvaluationObserver(ABC):
    @abstractmethod
    def on_event(self, event: dict[str, Any]) -> None: ...


class JsonLineObserver(EvaluationObserver):

    def __init__(self, stream: TextIO = sys.stderr) -> None:
        self._stream = stream

    def on_event(self, event: dict[str, Any]) -> None:
        self._stream.write(json.dumps(event, default=_json_default) + "\n")


class ConsoleObserver(EvaluationObserver):

    def __init__(self, level: str = "all") -> None:
        self._level = level

    def on_event(self, event: dict[str, Any]) -> None:
        if self._level == "deny_only" and event.get("type") == "evaluation.completed":
            result = event.get("result")
            if result is not None and getattr(result, "decision", None) is not None:
                if result.decision.value != "deny":
                    return
            elif isinstance(result, dict) and result.get("decision") != "deny":
                return
        print(f"[hushspec] {event.get('type')} at {event.get('timestamp')}", event, file=sys.stderr)


class MetricsCollector(EvaluationObserver):
    """Counters and a latency histogram over the evaluation stream.

    Safe to attach to a guard that several threads evaluate through: every
    counter update is taken under a lock. The duration samples are a bounded
    window (:data:`DURATION_WINDOW`), so the collector does not grow for the
    life of the process; the average and the percentile it reports describe
    that window, while the counters and the histogram are exact.

    :meth:`to_prometheus` renders the series the observability specification
    names, so a dashboard or a recording rule works against any SDK unchanged:

    ==============================  =========  =========================
    Series                          Type       Labels
    ==============================  =========  =========================
    ``hushspec_evaluate_total``     counter    ``decision``, ``action_type``
    ``hushspec_evaluate_duration_us``  histogram  ``action_type``
    ``hushspec_rule_match_total``   counter    ``rule_block``, ``decision``
    ``hushspec_policy_load_total``  counter    ``status``
    ==============================  =========  =========================
    """

    def __init__(self, duration_window: int = DURATION_WINDOW) -> None:
        self._lock = threading.Lock()
        self._counts: dict[str, int] = {}
        self._evaluations: dict[tuple[str, str], int] = {}
        self._rule_matches: dict[tuple[str, str], int] = {}
        self._policy_loads: dict[str, int] = {}
        self._histograms: dict[str, _Histogram] = {}
        self._durations: deque[float] = deque(maxlen=duration_window)
        self._evaluation_count = 0

    def on_event(self, event: dict[str, Any]) -> None:
        event_type = event.get("type", "")
        with self._lock:
            self._counts[event_type] = self._counts.get(event_type, 0) + 1
            if event_type == "evaluation.completed":
                self._record_evaluation(event)
            elif event_type in ("policy.loaded", "policy.reloaded"):
                self._policy_loads["success"] = self._policy_loads.get("success", 0) + 1
            elif event_type == "policy.load_failed":
                self._policy_loads["failure"] = self._policy_loads.get("failure", 0) + 1

    def _record_evaluation(self, event: dict[str, Any]) -> None:
        """Count one decision and record its latency. Called under the lock."""
        result = event.get("result")
        decision = "unknown"
        matched_rule: Optional[str] = None
        if isinstance(result, dict):
            decision = str(result.get("decision", "unknown"))
            matched = result.get("matched_rule")
            matched_rule = None if matched is None else str(matched)
        elif result is not None:
            decision = _decision_name(result.decision)
            matched_rule = result.matched_rule

        action = event.get("action")
        if isinstance(action, dict):
            action_type = str(action.get("type", ""))
        elif action is not None:
            action_type = action.type
        else:
            action_type = ""

        key = f"evaluate.{decision}"
        self._counts[key] = self._counts.get(key, 0) + 1
        self._evaluations[(decision, action_type)] = (
            self._evaluations.get((decision, action_type), 0) + 1
        )

        rule_block = rule_block_of(matched_rule)
        if rule_block is not None:
            self._rule_matches[(rule_block, decision)] = (
                self._rule_matches.get((rule_block, decision), 0) + 1
            )

        duration_us = max(0.0, float(event.get("duration_us", 0) or 0))
        histogram = self._histograms.setdefault(action_type, _Histogram())
        histogram.record(duration_us)
        self._durations.append(duration_us)
        self._evaluation_count += 1

    def get_count(self, key: str) -> int:
        with self._lock:
            return self._counts.get(key, 0)

    def get_total_evaluations(self) -> int:
        with self._lock:
            return self._evaluation_count

    def get_average_duration_us(self) -> float:
        """Mean latency over the sample window, in microseconds."""
        with self._lock:
            durations = list(self._durations)
        if not durations:
            return 0.0
        return sum(durations) / len(durations)

    def get_p99_duration_us(self) -> float:
        """99th percentile latency over the sample window, in microseconds."""
        with self._lock:
            durations = sorted(self._durations)
        if not durations:
            return 0.0
        index = min(int(len(durations) * 0.99), len(durations) - 1)
        return durations[index]

    def to_prometheus(self) -> str:
        """Prometheus text exposition (version 0.0.4) of every series."""
        with self._lock:
            evaluations = dict(self._evaluations)
            rule_matches = dict(self._rule_matches)
            policy_loads = dict(self._policy_loads)
            histograms = {name: histogram.copy() for name, histogram in self._histograms.items()}

        lines = [
            "# HELP hushspec_evaluate_total Total HushSpec evaluations",
            "# TYPE hushspec_evaluate_total counter",
        ]
        for (decision, action_type), count in sorted(evaluations.items()):
            lines.append(
                f'hushspec_evaluate_total{{decision="{decision}",'
                f'action_type="{_escape_label(action_type)}"}} {count}'
            )

        lines.append("# HELP hushspec_evaluate_duration_us Evaluation duration in microseconds")
        lines.append("# TYPE hushspec_evaluate_duration_us histogram")
        for action_type, histogram in sorted(histograms.items()):
            label = _escape_label(action_type)
            for bound, count in zip(DURATION_BUCKETS_US, histogram.buckets):
                lines.append(
                    f'hushspec_evaluate_duration_us_bucket{{action_type="{label}",'
                    f'le="{bound}"}} {count}'
                )
            lines.append(
                f'hushspec_evaluate_duration_us_bucket{{action_type="{label}",'
                f'le="+Inf"}} {histogram.count}'
            )
            lines.append(
                f'hushspec_evaluate_duration_us_sum{{action_type="{label}"}} '
                f"{_format_number(histogram.total)}"
            )
            lines.append(
                f'hushspec_evaluate_duration_us_count{{action_type="{label}"}} {histogram.count}'
            )

        lines.append("# HELP hushspec_rule_match_total Rule block match counts")
        lines.append("# TYPE hushspec_rule_match_total counter")
        for (rule_block, decision), count in sorted(rule_matches.items()):
            lines.append(
                f'hushspec_rule_match_total{{rule_block="{_escape_label(rule_block)}",'
                f'decision="{decision}"}} {count}'
            )

        lines.append("# HELP hushspec_policy_load_total Policy load operations")
        lines.append("# TYPE hushspec_policy_load_total counter")
        for status, count in sorted(policy_loads.items()):
            lines.append(f'hushspec_policy_load_total{{status="{status}"}} {count}')

        return "\n".join(lines) + "\n"

    def reset(self) -> None:
        with self._lock:
            self._counts.clear()
            self._evaluations.clear()
            self._rule_matches.clear()
            self._policy_loads.clear()
            self._histograms.clear()
            self._durations.clear()
            self._evaluation_count = 0


class ObservableEvaluator:
    """Evaluates a policy and announces the outcome to its observers."""

    def __init__(self, redact_content: bool = True) -> None:
        self._observers: list[EvaluationObserver] = []
        self._redact_content = redact_content

    def add_observer(self, observer: EvaluationObserver) -> None:
        self._observers.append(observer)

    def remove_observer(self, observer: EvaluationObserver) -> None:
        self._observers = [o for o in self._observers if o is not observer]

    def evaluate(self, spec: HushSpec, action: EvaluationAction) -> EvaluationResult:
        """Evaluate *action* against *spec* and emit ``evaluation.completed``.

        Routes through the detection pipeline, as every other evaluation
        surface does: for a policy carrying an ``extensions.detection`` block
        an escalated decision must not come back as an allow simply because
        the caller went through an observer (detection spec section 4).
        """
        from hushspec.detection import evaluate_with_detection

        start_ns = time.perf_counter_ns()
        result = evaluate_with_detection(spec, action).evaluation
        duration_us = (time.perf_counter_ns() - start_ns) // 1000
        self._emit({
            "type": "evaluation.completed",
            "timestamp": _iso_now(),
            "action": self._redact(action),
            "result": result,
            "duration_us": duration_us,
        })
        return result

    def notify_evaluation_completed(
        self,
        action: EvaluationAction,
        result: EvaluationResult,
        duration_us: int,
        enforcement: Optional["EnforcementSummary"] = None,
        receipt: Optional["DecisionReceipt"] = None,
    ) -> None:
        event: dict[str, Any] = {
            "type": "evaluation.completed",
            "timestamp": _iso_now(),
            "action": self._redact(action),
            "result": result,
            "duration_us": duration_us,
        }
        if enforcement is not None:
            event["enforcement"] = enforcement
        if receipt is not None:
            event["receipt"] = receipt
        self._emit(event)

    def _redact(self, action: EvaluationAction) -> EvaluationAction:
        """Return *action* with ``content`` stripped for observer emission.

        Evaluation itself (``evaluate()`` above) always runs against the real,
        unredacted action -- this only affects what gets embedded in observer
        events. A 0.2 receipt needs no such switch: it records the content's
        hash and size and has no field content could go in (receipt spec 4.4).
        """
        if self._redact_content and action.content is not None:
            return dataclasses.replace(action, content=None)
        return action

    def notify_policy_loaded(self, name: Optional[str] = None, hash: Optional[str] = None) -> None:
        """Announce the policy now in force. ``hash`` is its canonical content
        hash (``sha256:<hex>``), the same value its receipts carry."""
        self._emit({
            "type": "policy.loaded",
            "timestamp": _iso_now(),
            "policy_name": name,
            "content_hash": hash or "",
        })

    def notify_policy_load_failed(self, error: str, source: Optional[str] = None) -> None:
        self._emit({
            "type": "policy.load_failed",
            "timestamp": _iso_now(),
            "error": error,
            "source": source,
        })

    def notify_sink_error(self, error: str, source: Optional[str] = None) -> None:
        """Announce a receipt sink that refused what it was handed.

        The decision it belonged to stands: a sink is evidence, never
        enforcement. *source* names the sink that refused.
        """
        self._emit({
            "type": "sink.error",
            "timestamp": _iso_now(),
            "error": error,
            "source": source,
        })

    def notify_policy_reloaded(
        self,
        name: Optional[str] = None,
        hash: Optional[str] = None,
        previous_hash: Optional[str] = None,
    ) -> None:
        self._emit({
            "type": "policy.reloaded",
            "timestamp": _iso_now(),
            "policy_name": name,
            "content_hash": hash or "",
            "previous_hash": previous_hash,
        })

    def notify_error(self, error: str, source: Optional[str] = None) -> None:
        """Announce a failure the guard absorbed.

        Neither a policy load nor a sink: an observer that raised, say. The
        evaluation that produced the event stands -- an observer is a
        bystander, never enforcement.
        """
        self._emit({
            "type": "error",
            "timestamp": _iso_now(),
            "error": error,
            "source": source,
        })

    def _emit(self, event: dict[str, Any]) -> None:
        # Over a snapshot: an observer that registers another one while being
        # notified must not mutate the list this loop is walking.
        for observer in tuple(self._observers):
            try:
                observer.on_event(event)
            except Exception as exc:  # noqa: BLE001 - an observer is never fatal
                if event.get("type") == "error":
                    continue
                # The failure goes back to the observer that raised it: one
                # nobody is told about is the one that goes unnoticed. An
                # observer that raises reporting its own failure is dropped
                # rather than retried.
                try:
                    observer.on_event({
                        "type": "error",
                        "timestamp": _iso_now(),
                        "error": f"observer raised: {exc}",
                        "source": None,
                    })
                except Exception:  # noqa: BLE001
                    pass


def _decision_name(decision: Any) -> str:
    """A decision as the string a metric or a JSON line carries."""
    return decision.value if isinstance(decision, enum.Enum) else str(decision)


def _json_default(obj: Any) -> Any:
    from hushspec.receipt import DecisionReceipt, receipt_to_dict

    if isinstance(obj, DecisionReceipt):
        # Route through the shared helper (rather than a plain asdict) so a
        # receipt embedded in an observer event serializes identically to one
        # sent through a ReceiptSink: spec member order, enums as their string
        # values, and absent optional fields omitted rather than null.
        return receipt_to_dict(obj)
    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
        return dataclasses.asdict(obj)
    if isinstance(obj, enum.Enum):
        return obj.value
    raise TypeError(f"Object of type {type(obj).__name__} is not JSON serializable")


def _iso_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
